src/Compression/CompressionCodecQATZSTD.cpp

```diff
@@ -0,0 +1,114 @@
+#ifdef ENABLE_QATZSTD_COMPRESSION
+#include <Compression/CompressionInfo.h>
+#include <Compression/CompressionFactory.h>
+#include <zstd.h>
+#include <Parsers/IAST.h>
+#include <Parsers/ASTLiteral.h>
+#include <Parsers/ASTFunction.h>
+#include <Common/typeid_cast.h>
+#include <Poco/Logger.h>
+#include <Common/logger_useful.h>
+#include "qatseqprod.h"
+#include "CompressionCodecZSTD.h"
+
+namespace DB
+{
+namespace ErrorCodes
+{
+    extern const int CANNOT_COMPRESS;
+    extern const int ILLEGAL_SYNTAX_FOR_CODEC_TYPE;
+    extern const int ILLEGAL_CODEC_PARAMETER;
+}
+
+class CompressionCodecQATZSTD : public CompressionCodecZSTD
+{
+public:
+    static constexpr auto QATZSTD_SUPPORTED_MIN_LEVEL = 1;
+    static constexpr auto QATZSTD_SUPPORTED_MAX_LEVEL = 12;
+    explicit CompressionCodecQATZSTD(int level_);
+    ~CompressionCodecQATZSTD() override;
+
+protected:
+    UInt32 doCompressData(const char * source, UInt32 source_size, char * dest) const override;
+
+private:
+    const int level;
+    mutable bool initialized;
+    mutable ZSTD_CCtx* cctx;
+    mutable void *sequenceProducerState;
+    Poco::Logger * log;
+};
+
+UInt32 CompressionCodecQATZSTD::doCompressData(const char * source, UInt32 source_size, char * dest) const
+{
+    if (!initialized)
+    {
+        cctx = ZSTD_createCCtx();
```

Comment:
There is a risk that the initialization code (l. 44-62) runs more than once or in paralell. Can we create a constructor for `CompressionCodecQATZSTD` and move it there?

EDIT: At least we could try to move l. 48 and l. 50 into the constructor and store the `sequenceProducerState` as a class member variable. The rest of the initialization code can remain but I suppose we will no longer need the `if` in l. 44.

Reply:
Done